ngrok/src/config/policies.rs

```diff
@@ -0,0 +1,250 @@
+use std::{
+    borrow::Borrow,
+    fs::read_to_string,
+};
+
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use thiserror::Error;
+
+use crate::internals::proto;
+
+/// A set of policies that define rules that should be applied to incoming or outgoing
+/// connections to the edge.
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+#[serde(default)]
+pub struct Policies {
+    inbound: Vec<Policy>,
+    outbound: Vec<Policy>,
+}
+
+/// A policy that defines rules that should be applied
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+#[serde(default)]
+pub struct Policy {
+    name: String,
+    expressions: Vec<String>,
+    actions: Vec<Action>,
+}
+
+/// An action that should be taken if the policy matches
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+#[serde(default)]
+pub struct Action {
+    #[serde(rename = "type")]
+    type_: String,
+    config: Option<serde_json::Value>,
+}
+
+/// Error representing invalid string for Policies
+#[derive(Debug, Clone, Error)]
+#[error("invalid policies, err: {}", .0)]
+pub struct InvalidPolicies(String);
+
+/// Used just for json parsing
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+struct Envelope {
+    policies: Policies,
+}
+
+impl Policies {
+    /// Create a new empty [Policies] struct
+    pub fn new() -> Self {
+        Policies {
+            ..Default::default()
+        }
+    }
+
+    /// Create a new [Policies] from a json string
+    pub fn from_json(json: impl Into<String>) -> Result<Self, InvalidPolicies> {
+        let envelope: Envelope =
+            serde_json::from_str(&json.into()).map_err(|e| InvalidPolicies(e.to_string()))?;
+        Ok(envelope.policies)
+    }
+
+    /// Create a new [Policies] from a json file
+    pub fn from_file(json_file_path: impl Into<String>) -> Result<Self, InvalidPolicies> {
+        Policies::from_json(
+            read_to_string(json_file_path.into()).map_err(|e| InvalidPolicies(e.to_string()))?,
+        )
+    }
+
+    /// Convert [Policies] to json string
+    pub fn to_json(&self) -> Result<String, InvalidPolicies> {
+        let envelope = Envelope {
+            policies: self.clone(),
+        };
+        serde_json::to_string(&envelope).map_err(|e| InvalidPolicies(e.to_string()))
+    }
```

Comment:
@nijikokun do you have any feedback around this? It's in the product spec, and I've added the `to_json` for Policy and Action as talked about above. I would guess something like making it always possible to go back and forth from json to struct as a principle, regardless of how it may or may not be used initially?

Reply:
Oh, just realized that the fields of the `Policies` struct aren't `pub`, so users won't interact directly with the odd nesting. I guess that's fine then.

Just for our internal sanity, I would probably make a renamed `Envelope` the public type with the single `policies` field so that the `to/from_json` methods really don't add anything over the already-provided `Serialize/Deserialize` impls.

Reply:
Have now changed `Policies` to be the root, and made a `PolicySet` private layer to hold inbound/outbound.

Reply:
> I would guess something like making it always possible to go back and forth from json to struct as a principle, regardless of how it may or may not be used initially?

Correct, the idea here is to ensure that the configuration systems we have promote portability by supporting generation and accepting the same configuration. The generation portion is the reason for the `toJSON / to_json`.

A lot of this is assuming a few things:
- User will paste the configuration that is generated from the CLI or Dashboard into an SDK
- User will want to generate a configuration from the builder to put into the Cloud
- User will want to use the builder to pass to the SDK